types, rawdb, core, miner: add block location fields to receipt

bmperrea opened the pull request:
Solves https://github.com/ethereum/go-ethereum/issues/15210 without changing consensus, in a backwards compatible way, by adding the block location fields from ReadReceipt to the Receipt struct. 

Update: In particular this improves consistency between the go client interface and the [json rpc spec](https://github.com/ethereum/wiki/wiki/JSON-RPC#eth_gettransactionreceipt) and allows go clients to use transaction receipts to watch for block confirmations for particular transactions - see #15210.

bmperrea commented:
@karalabe @holiman any chance for a review on this 32 line pr?

bmperrea commented:
@karalabe @holiman are there any concerns with getting this through as it is now?

favadi commented:
@bmperrea probably they just had no time to do a review yet.

jgiles commented:
@karalabe any chance for a review on this?

holiman commented on core/types/gen_receipt_json.go:
```diff
@@ -47,6 +54,9 @@ func (r *Receipt) UnmarshalJSON(input []byte) error {
 		GasUsed           *hexutil.Uint64 `json:"gasUsed" gencodec:"required"`
+		BlockHash         *common.Hash    `json:"blockHash"`
+		BlockNumber       *hexutil.Big    `json:"blockNumber"`
+		TransactionIndex  *hexutil.Big    `json:"transactionIndex"`
```
Perhaps good to have `omitEmpty` on the json-part?

holiman commented on core/types/receipt.go:
```diff
@@ -55,13 +56,20 @@ type Receipt struct {
 	CumulativeGasUsed hexutil.Uint64
 	GasUsed           hexutil.Uint64
+	BlockNumber       *hexutil.Big
+	TransactionIndex  *hexutil.Big
```
I'm not too certain about this -- doesn't this mean that you change the RLP-representation?

bmperrea replied:
[`// receiptRLP is the consensus encoding of a receipt.`](https://github.com/ethereum/go-ethereum/pull/17662/files/5d0009093b1dd601069c18d6f4defc1547559ea0#diff-cf3bf12023ca997cd6b43322a70ea69dR75) so `receiptMarshaling` is just for struct->json.

fjl replied:
I can conform that receiptMarshaling just exists for JSON purposes.

bmperrea commented:
Thanks for the helpful review @holiman ! Do you still want to wait on a response from @fjl or @karalabe ?

holiman commented:
That would be good, yes, ping @fjl

bmperrea commented:
@karalabe @fjl @holiman Could we get a decisive review on this one? Thanks!

bmperrea commented:
thanks @karalabe :thumbsup:. Is the rpc spec you're referring to documented somewhere? I see block location fields are included here https://github.com/ethereum/wiki/wiki/JSON-RPC#eth_gettransactionreceipt

bmperrea commented:
Great point about mined and downloaded receipts @karalabe . I'm excited to improve the geth go-client interface to make sure we keep consistent with the rpc spec which includes these location fields. I've updated the PR to add code to take care of these cases in `miner/worker.go` and `core/blockchain.go` respectively. These are the same place where location-based fields are added to the log objects within the receipt. 

This pr makes the receipt object in the go-client interface more consistent with the rpc spec while also keeping consistent with how location fields are set for log entries.

bmperrea commented:
In general there are significant differences between the rpc api and the `ethclient` go client api. For instance 
`TransactionByHash(ctx context.Context, txHash common.Hash) (tx *types.Transaction, isPending bool, err error)` of the go client (`interfaces.go`) vs
`func (s *PublicTransactionPoolAPI) GetTransactionByHash(ctx context.Context, hash common.Hash) *RPCTransaction` (`ethapi/api.go`)

@karalabe is right that these missing block location fields, as well as the differences between `RPCTransaction` and `types.Transaction` are a problem. They are fragmenting the ecosystem because we are missing information that other clients report. 

This PR is only the beginning of fixing the broad issue.
We should put together and EIP to discuss the best way to match the `ethclient` interface to the RPC spec. One long-term option would be to make the json rpc a serialized version of the same struct returned by the `ethclient`.

rjl493456442 commented:
Hi, actually I don't think we need to add these additional fields into the `receipt` struct. We can retrieve these data in `txLookup` structure. Maybe we need this additional information in the RPC side, we can just retrieve these from the database and assemble the response.

And actually what we are doing now is "try to cut some unnecessary fields" to save disk usage. Like in your PR you add three new fields which occupy 48bytes for each receipt. It can abuse the disk storage. For more information check #17106

holiman commented:
We'll add this after https://github.com/ethereum/go-ethereum/pull/17106 is in

bmperrea commented:
Thanks @holiman @karalabe . I rebased and added the block location fields in `state_processor`.

To add the transaction index in `core/state_processor.go` I ended up exposing the current transaction index from the `statedb`. The other option would be to add that index as an argument to `statedb.ApplyTransaction` since each caller of that function needs to have the transaction index to call `statedb.Prepare` right before to setup that data. I think exposing from statedb is a reasonable approach given that the current pattern passes the transaction index in `Prepare` and the getter is not so different from other methods in `core/state/statedb.go`.

Good work @rjl493456442 on #17106 . The separation of Receipt struct from the corresponding rlp and db objects allows for storage optimization while still allowing us to add context that a client would expect to have available when a Receipt object is returned, and the same goes for other types returned by ethclient.

fjl commented:
This change is going to help with ethclient because it will make it possible to fetch inclusion info through TransactionReceipt

fjl commented on core/types/receipt.go:
```diff
@@ -55,13 +56,20 @@ type Receipt struct {
+	// Block location fields - not part of the rlp
+	BlockHash        common.Hash `json:"blockHash,omitempty"`
+	BlockNumber      *big.Int    `json:"blockNumber,omitempty"`
+	TransactionIndex *big.Int    `json:"transactionIndex,omitempty"`
```
`TransactionIndex` can just be a regular `uint`. Maybe it should be called `BlockIndex` to make it clear that it's the index in the block.

bmperrea replied:
I'll change the type 👍 

I got the name `TransactionIndex` from [this RPC spec](https://github.com/ethereum/wiki/wiki/JSON-RPC#eth_gettransactionreceipt). I'm hoping we can continue toward `ethclient` implementing that entire spec.

fjl replied:
Yeah, it can be called TransactionIndex, it's fine. ethclient will never implement the full RPC API because it is based on core/types and some of the concepts in the web3 RPC API cannot be represented using the types. One example is transaction sender and inclusion info: the RPC API returns inclusion info and sender in the transaction object, but it would be silly to add those fields to `types.Transaction`).

bmperrea replied:
I would rather see us separate the interface struct from the storage struct then to compromise on the interface. For example `ethclient` could return something more like what's in `rpcTransaction`. 

It would be good to have a go client that is as functional as rpc clients in other languages and I think `ethclient` should be that interface.

fjl replied:
I still want to get to the native implementation done. If you're seriously interested in improving Go developer experience using go-ethereum as a library, I'd be super happy if you could help us make that happen by implementing the existing interfaces on the server side

holiman commented on core/state_processor.go:
```diff
@@ -121,6 +121,9 @@ func ApplyTransaction(config *params.ChainConfig, bc ChainContext, author *commo
 	receipt.Bloom = types.CreateBloom(types.Receipts{receipt})
+	receipt.BlockHash = header.Hash()
+	receipt.BlockNumber = header.Number
+	receipt.TransactionIndex = uint(statedb.TxIndex())
```
A better solution might be to set them in Process, since you can reuse the hash instead of recalculating the header hash for every tx. 

The drawback, I guess, is that direct usage of `ApplytTransaction` won't get it, which I guess applies to the miner, the txpool (?) and various users of chain_makers in tests.

fjl replied:
@holiman fixed it now by adding an accessor for block hash to StateDB. The PR already adds an accessor for TxIndex, one more can't hurt.

holiman commented:
Lgtm!

zulhfreelancer commented:
Finally. Thank you guys!
